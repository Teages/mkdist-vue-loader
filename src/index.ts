import type { SFCBlock } from 'vue/compiler-sfc'
import type { InputFile, Loader, LoaderContext, LoaderResult, OutputFile } from './mkdist'

import fs from 'node:fs'
import { dirname, resolve } from 'pathe'

import { compileScript, parse } from 'vue/compiler-sfc'

export interface DefineVueLoaderOptions {
  blockLoaders?: {
    [blockType: string]: VueBlockLoader | undefined
  }
}

export interface VueBlockLoader {
  (
    block: Pick<SFCBlock, 'type' | 'content' | 'attrs'>,
    context: LoaderContext & {
      rawInput: InputFile
      blocks: SFCBlock[]
      addOutput: (...files: OutputFile[]) => void
    }
  ): Awaitable<Pick<SFCBlock, 'type' | 'content' | 'attrs'> | undefined>
}

export interface DefaultBlockLoaderOptions {
  type: 'script' | 'style' | 'template' | (string & {})
  outputLang: string
  defaultLang?: string
  validExtensions?: string[]
}

export function defineVueLoader(options?: DefineVueLoaderOptions): Loader {
  const blockLoaders = options?.blockLoaders || {}

  return async (input, context) => {
    if (input.extension !== '.vue') {
      return
    }

    const raw = await input.getContents()
    const sfc = parse(raw, { ignoreEmpty: true })

    if (sfc.errors.length > 0) {
      sfc.errors.forEach(error => console.error(error))
      return [{ path: input.path, contents: raw }]
    }

    const output: LoaderResult = []
    const addOutput = (...files: OutputFile[]) => output.push(...files)

    const blocks = [
      sfc.descriptor.template,
      ...sfc.descriptor.styles,
      ...sfc.descriptor.customBlocks,
    ].filter(item => !!item)
    // merge script blocks
    if (sfc.descriptor.script || sfc.descriptor.scriptSetup) {
      const merged = compileScript(sfc.descriptor, { id: input.path, fs: createFs(input) })
      merged.setup = false
      merged.attrs = toOmit(merged.attrs, 'setup')
      blocks.unshift(merged)
    }

    const results = await Promise.all(
      blocks.map(async (data) => {
        const blockLoader = blockLoaders[data.type]
        return await blockLoader?.(data, { ...context, rawInput: input, blocks, addOutput }) ?? data
      }),
    )

    const contents = results.map((block) => {
      const attrs = Object.entries(block.attrs)
        .map(
          ([key, value]) => {
            if (!value) {
              return undefined
            }

            return value === true ? key : `${key}="${value}"`
          },
        )
        .filter(item => !!item)
        .join(' ')

      const header = `<${(`${block.type} ${attrs}`).trim()}>`
      const footer = `</${block.type}>`

      return `${header}\n${cleanupBreakline(block.content)}\n${footer}\n`
    }).join('\n')
    addOutput({
      path: input.path,
      srcPath: input.srcPath,
      extension: '.vue',
      contents,
      declaration: false,
    })

    return output
  }
}

export function defineDefaultBlockLoader(options: DefaultBlockLoaderOptions): VueBlockLoader {
  return async (block, { loadFile, rawInput, addOutput }) => {
    if (options.type !== block.type) {
      return
    }

    const lang = typeof block.attrs.lang === 'string'
      ? block.attrs.lang
      : options.outputLang
    const extension = `.${lang}`

    const files
      = (await loadFile({
        getContents: () => block.content,
        path: `${rawInput.path}${extension}`,
        srcPath: `${rawInput.srcPath}${extension}`,
        extension,
      })) || []

    const blockOutputFile = files.find(
      f =>
        f.extension === `.${options.outputLang}`
        || options.validExtensions?.includes(f.extension as string),
    )
    if (!blockOutputFile?.contents) {
      return
    }
    addOutput(...files.filter(f => f !== blockOutputFile))

    return {
      type: block.type,
      attrs: toOmit(block.attrs, 'lang'),
      content: blockOutputFile.contents,
    }
  }
}

const styleLoader = defineDefaultBlockLoader({
  outputLang: 'css',
  type: 'style',
})

const scriptLoader = defineDefaultBlockLoader({
  outputLang: 'js',
  type: 'script',
  validExtensions: ['.js', '.mjs'],
})

export const vueLoader = defineVueLoader({
  blockLoaders: {
    style: styleLoader,
    script: scriptLoader,
  },
})

function createFs(input: InputFile) {
  const realpath = (...paths: string[]) => input.srcPath
    ? resolve(dirname(input.srcPath), ...paths)
    : resolve(...paths)
  const fileExists = (file: string) => {
    try {
      if (!input.srcPath) {
        return false
      }
      const path = realpath(file)

      fs.accessSync(path)
      return fs.lstatSync(path).isFile()
    }
    catch {
      return false
    }
  }
  const readFile = (file: string) => {
    return fs.readFileSync(realpath(file), 'utf-8')
  }

  return { realpath, fileExists, readFile }
}
function cleanupBreakline(str: string): string {
  return str.replaceAll(/(\n\n)\n+/g, '\n\n').replace(/^\s*\n|\n\s*$/g, '')
}
function toOmit<R extends Record<keyof object, unknown>, K extends keyof R>(record: R, toRemove: K): Omit<R, K> {
  return Object.fromEntries(Object.entries(record).filter(([key]) => key !== toRemove)) as Omit<R, K>
}

type Awaitable<T> = T | Promise<T>
