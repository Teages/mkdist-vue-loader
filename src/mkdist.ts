import type { MkdistOptions } from 'mkdist'

// The following types are copied from mkdist, they should be same to below types.

// export type Loader = Exclude<NonNullable<MkdistOptions['loaders']>[number], string>
// export type LoaderResult = Awaited<ReturnType<Loader>>
// export type InputFile = Parameters<Loader>[0]
// export type LoaderContext = Parameters<Loader>[1]

export type Loader = (input: InputFile, context: LoaderContext) => LoaderResult | Promise<LoaderResult>

type LoadFile = (input: InputFile) => LoaderResult | Promise<LoaderResult>

type LoaderOptions = Pick<
  MkdistOptions,
  | 'ext'
  | 'format'
  | 'declaration'
  | 'esbuild'
  | 'postcss'
>

export interface LoaderContext {
  loadFile: LoadFile
  options: LoaderOptions
}

export type LoaderResult = OutputFile[] | undefined

export interface InputFile {
  path: string
  extension: string
  srcPath?: string
  getContents: () => Promise<string> | string
}

export interface OutputFile {
  path: string
  srcPath?: string
  extension?: string
  contents?: string
  declaration?: boolean
  raw?: boolean
  skip?: boolean
}
