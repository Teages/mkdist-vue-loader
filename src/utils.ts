export function trimBreakLine(str: string): string {
  return str.replace(/\n$/, '').replace(/^\n/, '')
}

export type MaybePromise<T> = T | Promise<T>
