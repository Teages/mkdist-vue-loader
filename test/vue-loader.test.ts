import { readFile } from 'node:fs/promises'
import { resolve } from 'pathe'
import {
  beforeAll,
  describe,
  expect,
  it,
} from 'vitest'
import { vueLoader } from '../src'

describe('vue-loader', () => {
  let mkdist: typeof import('mkdist').mkdist

  beforeAll(async () => {
    mkdist = (await import('mkdist')).mkdist
  })

  it('vue-loader', async () => {
    const rootDir = resolve(__dirname, 'fixtures/vue-loader')
    const { writtenFiles } = await mkdist({ rootDir, loaders: ['js', 'postcss', 'sass', vueLoader], declaration: true })

    expect(writtenFiles.sort()).toEqual(
      [
        'dist/js-script-setup.vue',
        'dist/js-script-setup.vue.d.ts',
        'dist/js-script.vue',
        'dist/js-script.vue.d.ts',
        'dist/mixin.vue',
        'dist/mixin.vue.d.ts',
        'dist/ts-script-setup.vue',
        'dist/ts-script-setup.vue.d.ts',
        'dist/ts-script.vue',
        'dist/ts-script.vue.d.ts',
      ]
        .map(f => resolve(rootDir, f))
        .sort(),
    )

    const contents = await Promise.all(
      writtenFiles.map(filepath => readFile(filepath, 'utf-8')),
    )

    contents.forEach((file) => {
      expect(file).toMatchSnapshot()
    })
  })
})
