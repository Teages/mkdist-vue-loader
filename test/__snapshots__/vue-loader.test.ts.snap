// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-loader > vue-loader 1`] = `
"<script>
import { useModel as _useModel, mergeModels as _mergeModels } from 'vue'
import { computed, ref } from 'vue'


export default /*@__PURE__*/Object.assign({ name: 'JsScriptSetup' }, {
  props: /*@__PURE__*/_mergeModels({
  title: {
    type: String,
    required: true,
  },
  message: {
    type: String,
    default: 'Hello, Vue!',
  },
}, {
    "modelValue": { type: String, required: true },
    "modelModifiers": {},
  }),
  emits: ["update:modelValue"],
  setup(__props, { expose: __expose }) {
  __expose();





const value = _useModel(__props, "modelValue")

const count = ref(0)
function increment() {
  count.value++
}

const localValue = computed({
  get: () => value.value,
  set: val => value.value = val,
})

const __returned__ = { value, count, increment, localValue, computed, ref }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

})
</script>

<template>
  <div>
    <h1 :class="$style.title">
      {{ title }}
    </h1>
    <p>{{ message }}</p>
    <button @click="increment">
      Increment
    </button>
    <p>Count: {{ count }}</p>

    <slot name="custom-content" />

    <input v-model="localValue">
    <p>Local Value: {{ localValue }}</p>
  </div>
</template>

<style module>
.title{color:blue}
</style>
"
`;

exports[`vue-loader > vue-loader 2`] = `
"declare const _default: __VLS_WithTemplateSlots<typeof __VLS_component, __VLS_TemplateResult["slots"]>;
export default _default;
type __VLS_WithTemplateSlots<T, S> = T & (new () => {
    $slots: S;
});
declare const __VLS_component: import("vue").DefineComponent<{}, {
    title: string;
    message: string;
    $props: {
        readonly title?: string | undefined;
        readonly message?: string | undefined;
    };
}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<{}> & Readonly<{}>, {}, {}, {}, {}, string, import("vue").ComponentProvideOptions, true, {}, any>;
type __VLS_TemplateResult = {
    slots: {
        "custom-content"?(_: {}): any;
    };
    refs: {};
    attrs: Partial<typeof __VLS_inheritedAttrs>;
};
declare var __VLS_inheritedAttrs: {};
"
`;

exports[`vue-loader > vue-loader 3`] = `
"<script>
import { computed, defineComponent, ref } from 'vue'

function useCounter() {
  const count = ref(0)
  const increment = () => {
    count.value++
  }
  return { count, increment }
}

export default defineComponent({
  name: 'TsScript',

  props: {
    title: {
      type: String,
      required: true,
    },
    message: {
      type: String,
      default: 'Hello, Vue!',
    },
    modelValue: {
      type: String,
      default: '',
    },
  },

  emits: ['update:modelValue'],

  setup(props, { emit }) {
    const { count, increment } = useCounter()

    const localValue = computed({
      get: () => props.modelValue,
      set: value => emit('update:modelValue', value),
    })

    return {
      count,
      increment,
      localValue,
    }
  },
})
</script>

<template>
  <div>
    <h1>{{ title }}</h1>
    <p>{{ message }}</p>
    <button @click="increment">
      Increment
    </button>
    <p>Count: {{ count }}</p>

    <slot name="custom-content" />

    <input v-model="localValue">
    <p>Local Value: {{ localValue }}</p>
  </div>
</template>

<style scoped>
h1{color:blue}
</style>
"
`;

exports[`vue-loader > vue-loader 4`] = `
"declare const _default: import("vue").DefineComponent<import("vue").ExtractPropTypes<{
    title: {
        type: StringConstructor;
        required: true;
    };
    message: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: StringConstructor;
        default: string;
    };
}>, {
    count: import("vue").Ref<number, number>;
    increment: () => void;
    localValue: import("vue").WritableComputedRef<string, string>;
}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, "update:modelValue"[], "update:modelValue", import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{
    title: {
        type: StringConstructor;
        required: true;
    };
    message: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: StringConstructor;
        default: string;
    };
}>> & Readonly<{
    "onUpdate:modelValue"?: ((...args: any[]) => any) | undefined;
}>, {
    message: string;
    modelValue: string;
}, {}, {}, {}, string, import("vue").ComponentProvideOptions, true, {}, any>;
export default _default;
"
`;

exports[`vue-loader > vue-loader 5`] = `
"<script>
import { useModel as _useModel, mergeModels as _mergeModels, defineComponent as _defineComponent } from "vue";
import { computed, defineComponent, ref } from "vue";
const __default__ = defineComponent({ name: "TsScriptSetup" });
export default /* @__PURE__ */ _defineComponent({
  ...__default__,
  props: /* @__PURE__ */ _mergeModels({
    title: { type: String, required: true },
    message: { type: String, required: false }
  }, {
    "modelValue": { type: String, ...{ required: true } },
    "modelModifiers": {}
  }),
  emits: ["update:modelValue"],
  setup(__props, { expose: __expose }) {
    __expose();
    const value = _useModel(__props, "modelValue");
    const count = ref(0);
    function increment() {
      count.value++;
    }
    const localValue = computed({
      get: () => value.value,
      set: (val) => value.value = val
    });
    const __returned__ = { value, count, increment, localValue };
    Object.defineProperty(__returned__, "__isScriptSetup", { enumerable: false, value: true });
    return __returned__;
  }
});
</script>

<template>
  <div>
    <h1 :class="$style.title">
      {{ title }}
    </h1>
    <p>{{ message }}</p>
    <button @click="increment">
      Increment
    </button>
    <p>Count: {{ count }}</p>

    <slot name="custom-content" />

    <input v-model="localValue">
    <p>Local Value: {{ localValue }}</p>
  </div>
</template>

<style module>
.title{color:blue}
</style>
"
`;

exports[`vue-loader > vue-loader 6`] = `
"declare const _default: __VLS_WithTemplateSlots<import("vue").DefineComponent<import("vue").ExtractPropTypes<__VLS_TypePropsToOption<{
    modelValue: string;
} & {
    title: string;
    message?: string;
}>>, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    "update:modelValue": (modelValue: string) => void;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<__VLS_TypePropsToOption<{
    modelValue: string;
} & {
    title: string;
    message?: string;
}>>> & Readonly<{
    "onUpdate:modelValue"?: ((modelValue: string) => any) | undefined;
}>, {}, {}, {}, {}, string, import("vue").ComponentProvideOptions, true, {}, any>, {
    "custom-content"?(_: {}): any;
}>;
export default _default;
type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;
type __VLS_TypePropsToOption<T> = {
    [K in keyof T]-?: {} extends Pick<T, K> ? {
        type: import('vue').PropType<__VLS_NonUndefinedable<T[K]>>;
    } : {
        type: import('vue').PropType<T[K]>;
        required: true;
    };
};
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-loader > vue-loader 7`] = `
"<script>
import { useModel as _useModel, mergeModels as _mergeModels, defineComponent as _defineComponent } from "vue";
import { computed, ref } from "vue";
export default /* @__PURE__ */ _defineComponent({
  ...{ name: "TsScriptSetup" },
  props: /* @__PURE__ */ _mergeModels({
    title: { type: String, required: true },
    message: { type: String, required: false }
  }, {
    "modelValue": { type: String, ...{ required: true } },
    "modelModifiers": {}
  }),
  emits: ["update:modelValue"],
  setup(__props, { expose: __expose }) {
    __expose();
    const value = _useModel(__props, "modelValue");
    const count = ref(0);
    function increment() {
      count.value++;
    }
    const localValue = computed({
      get: () => value.value,
      set: (val) => value.value = val
    });
    const __returned__ = { value, count, increment, localValue };
    Object.defineProperty(__returned__, "__isScriptSetup", { enumerable: false, value: true });
    return __returned__;
  }
});
</script>

<template>
  <div>
    <h1 :class="$style.title">
      {{ title }}
    </h1>
    <p>{{ message }}</p>
    <button @click="increment">
      Increment
    </button>
    <p>Count: {{ count }}</p>

    <slot name="custom-content" />

    <input v-model="localValue">
    <p>Local Value: {{ localValue }}</p>
  </div>
</template>

<style module>
.title{color:blue}
</style>
"
`;

exports[`vue-loader > vue-loader 8`] = `
"declare let __VLS_typeProps: {
    title: string;
    message?: string;
};
type __VLS_PublicProps = {
    modelValue: string;
} & typeof __VLS_typeProps;
declare function __VLS_template(): {
    slots: {
        "custom-content"?(_: {}): any;
    };
    refs: {};
    attrs: Partial<{}>;
};
type __VLS_TemplateResult = ReturnType<typeof __VLS_template>;
declare const __VLS_component: import("vue").DefineComponent<import("vue").ExtractPropTypes<__VLS_TypePropsToOption<__VLS_PublicProps>>, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {
    "update:modelValue": (modelValue: string) => void;
}, string, import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<__VLS_TypePropsToOption<__VLS_PublicProps>>> & Readonly<{
    "onUpdate:modelValue"?: ((modelValue: string) => any) | undefined;
}>, {}, {}, {}, {}, string, import("vue").ComponentProvideOptions, true, {}, any>;
declare const _default: __VLS_WithTemplateSlots<typeof __VLS_component, __VLS_TemplateResult["slots"]>;
export default _default;
type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;
type __VLS_TypePropsToOption<T> = {
    [K in keyof T]-?: {} extends Pick<T, K> ? {
        type: import('vue').PropType<__VLS_NonUndefinedable<T[K]>>;
    } : {
        type: import('vue').PropType<T[K]>;
        required: true;
    };
};
type __VLS_WithTemplateSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
"
`;

exports[`vue-loader > vue-loader 9`] = `
"<script>
import { computed, defineComponent, ref } from "vue";
function useCounter() {
  const count = ref(0);
  const increment = () => {
    count.value++;
  };
  return { count, increment };
}
export default defineComponent({
  name: "TsScript",
  props: {
    title: {
      type: String,
      required: true
    },
    message: {
      type: String,
      default: "Hello, Vue!"
    },
    modelValue: {
      type: String,
      default: ""
    }
  },
  emits: ["update:modelValue"],
  setup(props, { emit }) {
    const { count, increment } = useCounter();
    const localValue = computed({
      get: () => props.modelValue,
      set: (value) => emit("update:modelValue", value)
    });
    return {
      count,
      increment,
      localValue
    };
  }
});
</script>

<template>
  <div>
    <h1>{{ title }}</h1>
    <p>{{ message }}</p>
    <button @click="increment">
      Increment
    </button>
    <p>Count: {{ count }}</p>

    <slot name="custom-content" />

    <input v-model="localValue">
    <p>Local Value: {{ localValue }}</p>
  </div>
</template>

<style scoped>
h1{color:blue}
</style>
"
`;

exports[`vue-loader > vue-loader 10`] = `
"declare const _default: import("vue").DefineComponent<import("vue").ExtractPropTypes<{
    title: {
        type: StringConstructor;
        required: true;
    };
    message: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: StringConstructor;
        default: string;
    };
}>, {
    count: import("vue").Ref<number, number>;
    increment: () => void;
    localValue: import("vue").WritableComputedRef<string, string>;
}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, "update:modelValue"[], "update:modelValue", import("vue").PublicProps, Readonly<import("vue").ExtractPropTypes<{
    title: {
        type: StringConstructor;
        required: true;
    };
    message: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: StringConstructor;
        default: string;
    };
}>> & Readonly<{
    "onUpdate:modelValue"?: ((...args: any[]) => any) | undefined;
}>, {
    message: string;
    modelValue: string;
}, {}, {}, {}, string, import("vue").ComponentProvideOptions, true, {}, any>;
export default _default;
"
`;
